### PR TITLE
Issue #3177 Head-Hold Yaw Without User Input

### PR DESCRIPTION
**Pull request by KenImhof**

### CODE REVIEW

File: src/main/flight/pid.c [hunk 1]
```diff
@@ -719,7 +719,8 @@ void pidController(void)
 {
     uint8_t headingHoldState = getHeadingHoldState();
 
-    if (headingHoldState == HEADING_HOLD_UPDATE_HEADING) {
+   //if (headingHoldState == HEADING_HOLD_UPDATE_HEADING) {
+    if (headingHoldState == HEADING_HOLD_UPDATE_HEADING || headingHoldState == HEADING_HOLD_DISABLED) { //Issue #3177 Heading Hold Yaw Without RC Input
```
Comment (digitalentity) [inline src/main/flight/pid.c]: This is not a proper fix. State `HEADING_HOLD_DISABLE` means that quad is tilted too much and heading control is risky. With proposed change any hard bank or pitch will reset the heading hold target which may lead to drift.

Proper fix would be to change `getHeadingHoldState()` function and modify the logic of detecting `HEADING_HOLD_DISABLED`.

Comment (KenImhof) [reply src/main/flight/pid.c]: Yes, but this fix only applies to recording the yaw current position.  Heading-hold still does not get applied if there is a significant angle.  The problem is when there is a significant angle, we stop updating the current compass position.  If the yaw input is then centered and then the quad levels, heading-hold kicks in and causes the quad to yaw back to the last position before its significant angle.

Comment (digitalentity) [reply src/main/flight/pid.c]: Problem is that if tilt is large you should not apply heading hold but you must keep the desired heading if yaw stick is centered. With your fix any sufficient tilt will reset the heading hold target regardless of yaw stick which is not the desired behavior.

Comment (digitalentity) [reply src/main/flight/pid.c]: Current behavior is to record a new heading hold target is yaw stick is not centered AND tilt is small. This is flawed and needs fixing. 

With your fix we'll record a new heading hold target when yaw stick is not centered OR tilt is large.

Proper logic would be to record a new heading hold target if yaw stick is not centered regardless of tilt.

Comment (KenImhof) [reply src/main/flight/pid.c]: Got it, so we should just update  the get HeadingHoldState:

static uint8_t getHeadingHoldState(void)
{
    // Don't apply heading hold if overall tilt is greater than maximum angle inclination
    if (calculateCosTiltAngle() < headingHoldCosZLimit) {
        if (ABS(rcCommand[YAW]) != 0 && FLIGHT_MODE(HEADING_MODE)) { //Issue #3177 Heading Hold Yaw Without RC Input
            return HEADING_HOLD_UPDATE_HEADING;
        }
        else {
            return HEADING_HOLD_DISABLED;
        }
    }

**Comment by digitalentity** [comment]
@KenImhof why did you close this? Are you going to open a new PR?

**Comment by KenImhof** [comment]
I did not mean to close the PR or this issue.  I deleted the project from my github repository so I could make the changes and submit a new PR. Not sure if this automatically closed the PR.
 
I also just got back from the field and tried the following code update, which did not work:

static uint8_t getHeadingHoldState(void)
{
    //Don't apply heading hold if overall tilt is greater than maximum angle inclination
    if (calculateCosTiltAngle() < headingHoldCosZLimit) {
        if (ABS(rcCommand[YAW]) != 0 && FLIGHT_MODE(HEADING_MODE)) { //Issue #3177 Heading Hold Yaw Without RC Input
            return HEADING_HOLD_UPDATE_HEADING;
        }
        else {
            return HEADING_HOLD_DISABLED;
        }
    }
..................................

With this code, if I flip the quad and hold a little rc_yaw just before the quad changes from right-side-up to up-side-down, when the quad levels and the heading-hold activates, the quad rotates back to the original position before the 180% compass flip.

**Comment by KenImhof** [comment]
I see what is happening.  Heading-hold is disabled when inverted.  I am going to try enabling heading-hold when inverted and only disable heading-hold during the large inclination angles.  I am going to try the following methods to see what works best to reverse the inverted headingHoldTarget:

1.       Update the heading hold target during yaw (current operation) and during the large inclination angle. I will also increase the Mag Update frequency to make sure it is adjusted during the inverted large inclination angle.

2.       Always store the headingHoldTarget as downward thrust and adjust the offset for reverse thrust during the heading-hold operations.

3.       If angle error is larger than x (possible 15-degrees), update headingHoldTarget to the current value.

 

Thanks,

**Comment by digitalentity** [comment]
@KenImhof great, thanks for looking into this

**Pull request closed without merging**